### ISSUE TITLE
VM Option -XX:hashCode=2 not working as expected

### ISSUE THREAD

**Issue by guekho64** [eclipse/openj9#4154]
I understand that every JVM implementation has different options, and that each one of them may or may not add compatibility for every option that any other JVM has, aside from the standard ones that are reglamentary, like -help.

That being said, I'd like to know if Eclipse OpenJ9 has support for using the _-XX:hashCode_ option, and if that's the case, what could be causing the JVM to not preserve hash codes between different executions of a program.

I admit my scenario it's quite odd, as I use this option to mantain consistency of the output from Object#hashCode between executions of a determined program, being able to store the results for later use.

Hotspot JVM outputs the desired results, as expected, but Eclipse J9's won't mantaint consistency between executions.

I've already read the docs, but they don't mention anything about _-XX:hashCode_. I went searching into the source code hosted in this repo, trying to find something about this, with no success.

Also I've alredy added some additional java options to ensure _-XX:hashCode_ isn't being ignored by the JVM, being those:

-XX:hashCode=2 -XX:+PositiveIdentityHash -XX:-IgnoreUnrecognizedVMOptions -XX:+UnlockExperimentalVMOptions

Is Eclipse OpenJ9 able to reproduce Hotspot JVM behavior? Or am I searching something unfindable?

**Comment by DanHeidinga** [comment]
OpenJ9 doesn't support the `-XX:hashCode` option.

> I use this option to mantain consistency of the output from Object#hashCode between executions of a determined program, being able to store the results for later use.

Can you expand on what it does and why it's important to your use case?  Would a JVMTI agent be a more portable choice to control the Object#hashCode value?

**Comment by pdbain-ibm** [comment]
OpenJ9's identity hashcode is based on the location of an object in the heap, which is non-deterministic and will generally vary from run to run.  Are you calling `theObject.hashCode()` or `System.identityHashCode(theObject)`?

**Comment by JasonFengJ9** [comment]
@guekho64 Hotspot JVM `-XX:hashCode=2` appears just return a constant `1` for most objects `hashCode()` except `String` objects from my quick experiments. So if your target objects are not `String`, `Object1.hashCode()` results could be hardcoded as `1` or just be ignored. If `String` objects have to be handled, my thought is to use the `String` values directly just like package name and other staff you mentioned.
With these modification, I hope your program can adopt `OpenJ9 JVM` and also doesn't require `-XX:hashCode=2 -XX:+UnlockExperimentalVMOptions`.

**Comment by JasonFengJ9** [comment]
@guekho64 what Hotspot JVM version/platform do you run against the program?

**Comment by guekho64** [comment]
@JasonFengJ9 sorry for the late response. Here's what `java -version` outputs:

```
openjdk version "1.7.0_201"
OpenJDK Runtime Environment (Zulu 7.25.0.5-linux64) (build 1.7.0_201-b02)
OpenJDK 64-Bit Server VM (Zulu 7.25.0.5-linux64) (build 24.201-b02, mixed mode)
```

Although I've been using different versions of Hotspot JVM's since I started with this project of mine, including:

* OpenJDK 7/8
* Zulu JDK 7/8
* OpenJDK 8 (w/Shenandoah GC Patches)
* Oracle's JDK 7/8
* OpenJDK 8 (w/Eclipse J9's JVM)

And I haven't seen nothing out of usual when running my program with the discussed JVM option...

In fact, the program runs even better with Eclipse J9's JVM than with Hotspot, but there's this little detail that keeps me from using it for my project: The lack of support from this JVM for using the `-XX:hashCode` option, or a way to force the JVM to *not* scramble hashCodes every time the program is executed.

I know that there are many reasons to not preserve hash codes from execution to execution, like security and alike, and as such, I see 3 possibilities to solve this particular issue:

* Include support for `-XX:hashCode` JVM option, so user can decide whether or not hashCodes should change. (Like Hotspot does)

* Provide patches for users who might want this functionality...

And the last one:

* If none of the above is viable, maybe just giving out some clues on how to modify the source code should suffice, so people could just create custom builds of the JVM

Anyways, this is a personal project, so, in case that Eclipse OpenJ9 developers reject having official support for this feature, I can deal with it, just having a custom build of this JVM would be enough for me...although It may be useful for someone else, but I recognize that not always is that easy to add random features requested by people...

Another option could be to go back to Hotspot, but It'll be hard now that I know about this new virtual machine implementation, _sigh..._

**Comment by JasonFengJ9** [comment]
@guekho64 glad to hear that "In fact, the program runs even better with Eclipse J9's JVM than with Hotspot".
To preserve hash codes from execution to execution, a quick thought is to provide `java.util. Random` alike mechanism to generate pseudorandom numbers within `java.lang.Object.hashCode()` via natives.
Will let you know when such a patch is available for test.

**Comment by JasonFengJ9** [comment]
@guekho64 considering this is your personal project, I would propose a Java only patch to generate hash codes via `java.util.Random` within `java.lang.Object`. This function can be enabled via a system property `preserveHashCode`.

Important note: This patch doesn't apply to `String` objects which are handled differently by natives. In addition footprint is going to be increased and the performance might be affected as well.

If your program can't tolerate those limitations, a native implementation need to be investigated instead.

Attached jar file is for `OpenJ9 Java 8` and can be applied via `-Xbootclasspath/p:/path/to/patch-openj9-java8.zip` along with `-DpreserveHashCode=true`. 
[patch-openj9-java8.zip](https://github.com/eclipse/openj9/files/2725495/patch-openj9-java8.zip)

**Comment by guekho64** [comment]
Thanks for putting your own effort into this!

I'm gonna try this patch as soon as possible, since I'm not in my developing environment right now, but I'll let you know if this worked at all or not.

How hard would it be to implement this behavior for `String`'s too? As I said before, I use some Strings alongside other factors to determine which Object is which, and they're really useful because they contain the name of the class who invoked the method, and those names have very little chances to be repeated by someone else (Unless it's done on purpose...), so you could say that this acts as a way to confirm (Even more) if an Object it's the one being search...

A little hit on performance shouldn't be something to worry about, unless it _really_ affects the overall speed.

If necessary, I offer you my help to try to implement this.

**Comment by JasonFengJ9** [comment]
@guekho64 no problem, it's our job to help out `OpenJ9` usages, :)
Just to clarify about `String` hashcode, actually it was calculated based on the `String` content hence its hashcode already is preserved from execution to execution even without this patch.
So please give this patch a try and see if it unblocks your program from adapting `OpenJ9 jvm`.
I don't believe the speed will be a big hit but you will find out if it is noticeable for your program.

From the specification perspective (https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Object.html#hashCode()), _This (hashcode) integer need not remain consistent from one execution of an application to another execution of the same application._ , preserving hashcode from execution to execution is not required by Java spec.

Note: I don't expect this patch to be accepted into `OpenJ9` repo due to the footprint increasing. A proper implementation would be made within natives.

`OpenJ9` project always welcome contributions, big or small. If you would like to take this implementation, I (and others) will help. Thanks for your interest.

**Comment by guekho64** [comment]
Just a few seconds ago I have just tested my program alongside the proposed patches and...it didn't work, unfortunately.

it didn't work wether I was asking for Objects or other custom types `hashCode()`. In the meantime I'm gonna try to search for the Hotspot implementation and try to adapt it here, if possible.

<details><summary>Maybe I should open another issue for this...</summary>
<br>
Also, ironically, because the program is way faster than before, a new problem has arised:

<pre>
Pointer: 00000000FFFFFFFF has is not object aligned (to 8 bytes) 
00:00:04.191 0x7fb3940b5900    j9mm.141    *   ** ASSERTION FAILED ** at ./MarkingScheme.hpp:70: ((false))
</pre>

**Brief Explanation:** This problem appears after the program exits sucessfully, and immediatly I run the program again, within the next 1 || 2 seconds, and appears at random phases of the program. Maybe due to the use of `-Xbootclasspath/p:path`? It's the first time I see this error.

I guess that, maybe, there's a race condition hidden somewhere? Or somehow the JVM didn't release the memory used in the last process, and the OS reassigns the same memory area to the new process, failing due to the unclean state of that specific area?


```
openjdk version "1.8.0_192"
OpenJDK Runtime Environment (build 1.8.0_192-b12_openj9)
Eclipse OpenJ9 VM (build openj9-0.11.0, JRE 1.8.0 Linux amd64-64-Bit 20181107_88 (JIT enabled, AOT enabled)
OpenJ9   - 090ff9dcd
OMR      - ea548a66
JCL      - b5a3affe73 based on jdk8u192-b12)
```

</details>

**Comment by JasonFengJ9** [comment]
@guekho64 any chance you could share a testcase to show that JVM doesn't preserve the hash codes from execution to execution?

Yes, please open another issue for the assertion failure.

Please be cautious about searching for Hotspot implementation, `OpenJ9` was (and still is) a clean room JVM implementation which means no code/algorithm can be copied from Hotspot.

**Comment by guekho64** [comment]
@JasonFengJ9

The next piece of code yields the next results:

<details><summary>Code</summary>

```java

package misc.testing;

import static java.lang.System.out;
import static misc.testing.Main.Local.Methods.getDate;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public final class Main {
    public static final class Local {
        public static final class Methods {
            public static final String getDate () {
                final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MM-dd-yyyy HH:mm:ss");
                return LocalDateTime.now().format(formatter);
            }
        }
        public static final class Types {
            public static final class CustomizedObject {
                private final int placeholderInteger;
                private final String placeholderString;
                private final Object placeholderObject;
                public CustomizedObject () {
                    super();
                    this.placeholderInteger = 10;
                    this.placeholderString = "string";
                    this.placeholderObject = new Object();
                }
            }
        }
    }
    public static final void main (final String... arguments) {
        final Object normalObject = new Object();
        final Local.Types.CustomizedObject customizedObject = new Local.Types.CustomizedObject();
        out.println(normalObject.hashCode());
        out.println(customizedObject.hashCode());
        out.print(getDate());
    }
}

```
</details>

<details><summary>Results:</summary>

```
1155869325
155629808
01-04-2019 22:04:34
```

```
431529176
1429008869
01-04-2019 22:08:28
```

```
155629808
1429008869
01-04-2019 22:08:44
```

```
1155869325
155629808
01-04-2019 22:09:22
```

These results were obtained using the patches alongside with:

```
openjdk version "1.8.0_192"
OpenJDK Runtime Environment (build 1.8.0_192-b12_openj9)
Eclipse OpenJ9 VM (build openj9-0.11.0, JRE 1.8.0 Linux amd64-64-Bit 20181107_88 (JIT enabled, AOT enabled)
OpenJ9   - 090ff9dcd
OMR      - ea548a66
JCL      - b5a3affe73 based on jdk8u192-b12)
```
</details>

**Comment by JasonFengJ9** [comment]
@guekho64 did you run the program with `-DpreserveHashCode=true`?

**Comment by JasonFengJ9** [comment]
@guekho64 I am not able build the original code snippet due to missing class `misc.testing.Main.Local.Methods.getDate`. Commenting out the class in question (along with its references), the hashcode output appears same from execution to execution as following:
```
openjdk version "1.8.0_192"
OpenJDK Runtime Environment (build 1.8.0_192-b12_openj9)
Eclipse OpenJ9 VM (build openj9-0.11.0, JRE 1.8.0 Linux amd64-64-Bit Compressed References 20181107_95 (JIT enabled, AOT enabled)
OpenJ9   - 090ff9dcd
OMR      - ea548a66
JCL      - b5a3affe73 based on jdk8u192-b12)

1915244828
226796111
```
Could you supply a testcase without dependency (maybe attach the `.class` file as well)?

**Comment by guekho64** [comment]
@JasonFengJ9

Wow, such a mess right here, heh. In fact, the "missing" class is just an inner class, as the code shows, and the "depency" is supplied within the code, so I don't undertand what's going wrong at all...and `getDate` isn't a class at all, it's a method, lol.

Maybe Eclipse is doing something weird when compiling the code? 'Cause I use that IDE for my tests. And it works just fine by copying & pasting the above snippet in a new class...

And yes, I was using the new `-DpreserveHashCode=true` option while doing my tests, so this situation is quite...odd.

Fortunately, for odd situations, odd solutions.

<details><summary>Solution?</summary>
<br>
As I was seeing that accomplishing this was way harder than I initially thought, then I came up with a new way to verify objects, without using hash codes at all. In fact, ignoring them completely:
<br>
<br>
Here's my workaround:
<br>
<br>
<pre>
Please note that whenever you add and/or remove an element from a `List`, and then extract a SHA-256 string from it, the resulting String will change, and if the same `List` is restored to a previous state, it'll also recover its old SHA-256 key.

SHA-256 strings also get influenced somehow if the name of the file is changed, but restoring previous file name will also restore previous results...
</pre>
[HashingTest.java](https://github.com/guekho64/PersonalStuff/blob/master/Misc/HashingTest.java)
[Hashingtest.jar](https://github.com/guekho64/PersonalStuff/blob/master/Misc/HashingTest.jar)

**Comment by JasonFengJ9** [comment]
@guekho64 just realize that I modified the package name when coping the testcase code which caused `The import misc.testing.Main cannot be resolved` error. I do apologize not checking the code thoroughly before assuming the _class is missing_. (Honestly I didn't expect it imports a `class/method` within itself).
Now I run the original testcase and it still works as expected in a few runs with following result along with JVM version:
```
openjdk version "1.8.0_192"
OpenJDK Runtime Environment (build 1.8.0_192-b12_openj9)
Eclipse OpenJ9 VM (build openj9-0.11.0, JRE 1.8.0 Linux amd64-64-Bit Compressed References 20181107_95 (JIT enabled, AOT enabled)
OpenJ9   - 090ff9dcd
OMR      - ea548a66
JCL      - b5a3affe73 based on jdk8u192-b12)

1915244828
226796111
```
Good to hear that you applied `-DpreserveHashCode=true` earlier, but it is not clear why your environment got inconsistent result.

Anyway, it appears you find a solution without relying on `Object.hashCode()`. Hope your program works well with `OpenJ9 JVM`.

P.S. It won't be a trivial implementation to preserve hashcode from execution to execution, since this function is not required by Java spec, it probably won't be a priority item either. 

So if you agree, can this issue be closed?

**Comment by guekho64** [comment]
@JasonFengJ9

"Good to hear that you applied -DpreserveHashCode=true earlier, but it's not clear why your environment got inconsistent results" ~ Yes, I thought the same. I have no idea why your patch didn't work as expected on my side...

In fact, that's why the code outputs the date each time is executed, to show that each time the code was run, it printed a totally new value. Maybe the answer will never be found...

<hr>

Anyways, fortunately I found a way to overcome my dependency on `-XX:hashCode=2`, and, for the sake of knowledge, I left a workaround within my last comment, so I hope nobody else has to depend on this kind of obscure VM options...

In fact, I did so because it was my last ressort while trying to solve my problem, but initially I didn't expect to migrate away from Hotspot...

**_Then let's close this issue once for all!_**

PS: I'm just curious if the hash printed by my test, in my environment, would be the same if a different computer executed (Or compiled & executed) the same code...

**Comment by guekho64** [comment]
Thanks for helping me solve this one, I really appreciate your time spent here. I hope that Eclipse J9 will continue growing and getting better (Even better than it alredy is).

See ya until the next issue.
